src/components/number/number.test.jsx

```diff
@@ -0,0 +1,188 @@
+import React, { createRef } from 'react';
+import {
+  render,
+  screen,
+  cleanup,
+} from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import Number from './number.jsx';
+
+describe('Number', () => {
+  const requiredProps = {
+    id: 'test-component',
+    label: 'Test Component',
+  };
+
+  afterEach(cleanup);
+
+  it('has defaults', () => {
+    const ref = createRef();
+    render(<Number {...requiredProps} ref={ref} />);
+    expect(document.body).toMatchSnapshot();
+    expect(ref.current).toMatchSnapshot();
+  });
+
+  describe('disabled API', () => {
+    it('permits disabling', () => {
+      render(<Number {...requiredProps} disabled />);
+      expect(screen.getByLabelText('Test Component')).toBeDisabled();
+    });
+  });
+
+  describe('label API', () => {
+    it('presents a label', () => {
+      render(<Number {...requiredProps} />);
+      expect(screen.getByLabelText('Test Component')).toBeInTheDocument();
+    });
+  });
+
+  describe('ref API', () => {
+    it('responds with validity from the DOM', () => {
+      const ref = createRef();
+      render(<Number {...requiredProps} value={1.01} ref={ref} />);
+      expect(ref.current.validity).toBe(screen.getByLabelText('Test Component').validity);
+    });
+
+    it('indicates if it is dirty or not', () => {
+      const ref = createRef();
+      render(<Number {...requiredProps} value={1} ref={ref} />);
+      expect(ref.current.dirty).toBe(false);
+      userEvent.type(screen.getByLabelText('Test Component'), '1');
+      expect(ref.current.dirty).toBe(true);
+    });
+
+    it('permits focusing', () => {
+      const ref = createRef();
+      render(<Number {...requiredProps} ref={ref} />);
+      expect(document.activeElement).not.toBe(screen.getByLabelText('Test Component'));
+      ref.current.focus();
+      expect(document.activeElement).toBe(screen.getByLabelText('Test Component'));
+    });
+  });
+
+  describe('name API', () => {
+    it('presents the name on the component', () => {
+      render(<Number {...requiredProps} name="foo-bar" />);
+      expect(screen.getByLabelText('Test Component')).toHaveAttribute('name', 'foo-bar');
+    });
+
+    it('exposes the name on the ref', () => {
+      const ref = createRef();
+      render(<Number {...requiredProps} name="foo-bar" ref={ref} />);
+      expect(ref.current.name).toEqual('foo-bar');
+    });
+  });
+
+  describe('step API', () => {
+    it('sets the step', () => {
+      render(<Number {...requiredProps} step={0.1} />);
+      expect(screen.getByLabelText('Test Component')).toHaveAttribute('step', '0.1');
+    });
+  });
+
+  describe('placeholder API', () => {
+    it('allows setting placeholder text', () => {
+      render(<Number {...requiredProps} placeholder="Hello!" />);
+      expect(screen.getByLabelText('Test Component')).toHaveAttribute('placeholder', 'Hello!');
+    });
+  });
+
+  describe('required API', () => {
+    it('can be set', () => {
```

Comment:
Let's add an assertion on the validation message. Something like:

```
expect(...).toHaveDescription(...);
```

It tests that the input and the field control components are synced.